### PR TITLE
Fix: Fixed an issue where the title bar drag region was sometimes missing

### LINKED ISSUE #14958: Bug: Drag gap between tab bar and windows controls doesn't always happen
### Description

Currently sometimes show like this
image

### Steps To Reproduce

1. Tested on resolution 1920x1080 and scaling 100%
2. Maximize window
3. Open new tabs until tab bar is full

### Requirements

- Make sure a drag gap is always present between the tabs and window controls

### Files Version

3.2.7.0

### Windows Version

10.0.22621.3155

### Log File

[debug.log](https://github.com/files-community/Files/files/14591229/debug.log)

### LINKED ISSUE #16532: Bug:
### Description

I found a problem that appears in all versions of Files.
It is a problem with not keeping the distance after the application starts if more tabs are open than the width of the window.
Image

### Steps To Reproduce

Ensuring the correct padding distance

### Files Version

3.8.0.0

### Windows Version

10.0.26100.2314

### User ID

e7721cff-79b9-4d35-941d-e59b2b4fcc91

### Log File

There is no need

### PR DESCRIPTION
**Pull request by XTorLukas**
**Resolved / Related Issues**

To prevent extra work, all changes to the Files codebase must link to an approved issue marked as `Ready to build`. Please insert the issue number following the hashtag with the issue number that this Pull Request resolves.
- Closes #16532
- Closes #14958

**Steps used to test these changes**

Stability is a top priority for Files and all changes are required to go through testing before being merged into the repo. Please include a list of steps that you used to test this PR.

1. Open enough tabs that the scroll buttons are displayed
2. Fully close Files (enable continue where you left off)
3. Reopen Files
4. Tabs don't overflow and the gap is maintained.

image

- [x] Gap is identical at different dpi's e.g. 1.25

### CODE REVIEW

File: src/Files.App/Views/MainPage.xaml.cs [hunk 1]
```diff
@@ -125,8 +125,10 @@ private void UserSettingsService_OnSettingChangedEvent(object? sender, SettingCh
 		}
 
 		private void HorizontalMultitaskingControl_Loaded(object sender, RoutedEventArgs e)
+			=> TabControl.TabControlAreaSizeChanged += TabControlArea_SizeChanged;
+
+		private void TabControlArea_SizeChanged(object sender, SizeChangedEventArgs e)
```
Comment (0x5bfa) [inline src/Files.App/Views/MainPage.xaml.cs:133]: I don't think you should hook size event in here. This method hooks CurrentInstanceChanged every time the size changed. I think loaded event was good.

Comment (XTorLukas) [reply src/Files.App/Views/MainPage.xaml.cs:133]: Yea,  but I've already committed to a more elegant solution.

File: src/Files.App/MainWindow.xaml.cs [hunk 2]
```diff
@@ -339,5 +342,15 @@ x.tabItem.NavigationParameter.NavigationParameter is PaneNavigationArguments pan
 				}
 			}
 		}
+
+		public double GetDpiScale()
```
Comment (yaira2) [inline src/Files.App/MainWindow.xaml.cs:346]: We already have `AppWindowDPI` in AppModel, is there any way to reuse it?

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs [hunk 3]
```diff
@@ -28,6 +30,8 @@ public sealed partial class TabBar : BaseTabBar, INotifyPropertyChanged
 
 		private bool _lockDropOperation = false;
 
+		private static int _gap = 6;
```
Comment (yaira2) [inline src/Files.App/UserControls/TabBar/TabBar.xaml.cs:33]: We don't need to add an additional gap.

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs [hunk 4]
```diff
@@ -51,6 +55,14 @@ public bool AllowTabsDrag
 		public Rectangle DragArea
 			=> DragAreaRectangle;
 
+		private GridLength _titleBarWidth;
+
+		public GridLength TitleBarWidth
+		{
+			get => _titleBarWidth;
+			set => _titleBarWidth = new((value.Value / MainWindow.Instance.GetDpiScale()) + _gap);
```
Comment (yaira2) [inline src/Files.App/UserControls/TabBar/TabBar.xaml.cs:63]: Please see if you can use `App.AppModel.AppWindowDPI`.

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs (hunk 3 above)
Comment (yaira2) [inline src/Files.App/UserControls/TabBar/TabBar.xaml.cs:33]: Is this still needed?

File: src/Files.App/MainWindow.xaml.cs (hunk 2 above)
Comment (XTorLukas) [reply src/Files.App/MainWindow.xaml.cs]: Thanks, I didn't know.

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs (hunk 4 above)
Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs]: Thanks, I didn't know.

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs (hunk 3 above)
Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs:33]: Yes, I would still leave a gap, most programmes use it. On a touchscreen it's very useful to have a space to catch the window. But I would probably use a gap (minimum size) of 40px.

Comment (yaira2) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs]: I meant that there should already be the correct size gap from the inset properties.

Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs]: Without a gap, the space is exactly the size of TitleBar.Inset(Right or Left), depending on the culture.

> image

Green: TabBar.Inset
Red: minimum size `_gap`

Comment (yaira2) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs]: I'm seeing a different behavior on my side 
image

Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs]: it's the same, the size of TitleBar + _gap 🤔

Comment (yaira2) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs]: As you can see, the gap is already there, however if you follow the steps listed in #14598, the gap is not set correctly.

Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs]: > As you can see, the gap is usually there by default, however if you follow the steps listed in #14598, the gap is not set correctly.

For me so far the gap is always set correctly, I haven't found a problem yet

Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs]: If you mean this problem #14952 , I have already solved it in this PR

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs [hunk 5]
```diff
@@ -84,6 +89,40 @@ public TabBar()
 			};
 		}
 
+		/// <summary>
+		/// Initializes the title bar asynchronously. This method attempts to get the title bar's left and right insets.
+		/// If the values are properly initialized, it sets the <see cref="TitleBarWidth"/> property. If the initialization fails,
+		/// it falls back to a default width.
+		/// </summary>
+		private async void InitializeTitleBarAsync()
```
Comment (yaira2) [inline src/Files.App/UserControls/TabBar/TabBar.xaml.cs:97]: Do we still need multiple attempts now that you added `UIElement.Measure`?

Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs:97]: UIElement.Measure I added this before this test. But then you described a problem that never appeared during my testing. If this problem was only caused by a missing UIElement.Measure, then this test would no longer be necessary.

Comment (yaira2) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs:97]: Let's see what happens if we remove it, hopefully `UIElement.Measure` is all we need.

Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs:97]: In my personal testing, this has not manifested itself, but I plan to test this further on a weaker laptop where slower loading and thus the problem you describe may manifest itself.

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs [hunk 6]
```diff
@@ -51,6 +55,14 @@ public bool AllowTabsDrag
 		public Rectangle DragArea
 			=> DragAreaRectangle;
 
+		private GridLength _titleBarWidth;
+
+		public GridLength TitleBarWidth
```
Comment (yaira2) [inline src/Files.App/UserControls/TabBar/TabBar.xaml.cs:60]:
```suggestion
		public GridLength TitleBarRightInsetWidth
```

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs [hunk 7]
```diff
@@ -366,5 +374,25 @@ private void TabViewItem_Loaded(object sender, RoutedEventArgs e)
...
+			if (HorizontalTabView.ActualWidth > 0 && TabBarAddNewTabButton.Width > 0 && TitleBarWidth.Value > 0)
+			{
+				HorizontalTabView.Measure(new(HorizontalTabView.ActualWidth - TabBarAddNewTabButton.Width - TitleBarWidth.Value, HorizontalTabView.ActualHeight));
+				return;
+			}
+
+			DispatcherQueue.TryEnqueue(() => DragAreaRectangle_Loaded(sender, e));
+		}
+
+		public int SetTitleBarDragRegion(InputNonClientPointerSource source, SizeInt32 size, double scaleFactor, Func<UIElement, RectInt32?, RectInt32> getScaledRect)
```
Comment (yaira2) [inline src/Files.App/UserControls/TabBar/TabBar.xaml.cs:389]: Is it possible to move this method back to `MainPage`?

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs [hunk 8]
```diff
@@ -51,6 +55,14 @@ public bool AllowTabsDrag
 		public Rectangle DragArea
 			=> DragAreaRectangle;
 
+		private GridLength _titleBarWidth;
+
+		public GridLength TitleBarWidth
+		{
+			get => _titleBarWidth;
+			set => _titleBarWidth = new((value.Value / App.AppModel.AppWindowDPI) + _gap);
```
Comment (yaira2) [inline src/Files.App/UserControls/TabBar/TabBar.xaml.cs:63]: Assuming the `TitleBar.Inset` property isn't null, I think it's already DPI aware and we can use the value as is.

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs (hunk 6 above)
Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs:60]: It is possible to change, but a gap is already added to the value.

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs (hunk 7 above)
Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs:389]: @0x5bfa  suggested to convert all TabBar operations into one file, and for me it is reasonable

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs (hunk 8 above)
Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs:63]: No, I tested with 1.25 dpi and the value is not recalculated.

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs (hunk 7 above)
Comment (yaira2) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs:389]: I don't think we want to restrict the logic to the TabBar, this way all our options are open if we ever add a scenario where the TabBar is hidden.

Comment (XTorLukas) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs:389]: I don't know, my preference is to keep the changes

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs [hunk 9]
```diff
@@ -366,5 +359,39 @@ private void TabViewItem_Loaded(object sender, RoutedEventArgs e)
...
+					titleBarInset += 40; // Add 40px gap
+					RightPaddingColumn.Width = new(titleBarInset);
+
+					HorizontalTabView.Measure(new(
+						HorizontalTabView.ActualWidth - TabBarAddNewTabButton.Width - titleBarInset,
+						HorizontalTabView.ActualHeight));
+
+					return;
+				}
+			}
```
Comment (yaira2) [inline src/Files.App/UserControls/TabBar/TabBar.xaml.cs:383]:
```suggestion
		private void DragAreaRectangle_Loaded(object sender, RoutedEventArgs e)
		{
			double scaleAdjustment = DragAreaRectangle.XamlRoot.RasterizationScale;
    		double titleBarInset = ((FilePropertiesHelpers.FlowDirectionSettingIsRightToLeft
        		? MainWindow.Instance.AppWindow.TitleBar.LeftInset
        		: MainWindow.Instance.AppWindow.TitleBar.RightInset) / scaleAdjustment) + 48;

   			HorizontalTabView.Measure(new(HorizontalTabView.ActualWidth - TabBarAddNewTabButton.Width - titleBarInset, HorizontalTabView.ActualHeight));
    		RightPaddingColumn.Width = new(titleBarInset >= 0 ? titleBarInset : 0);
		}
```

File: src/Files.App/UserControls/TabBar/TabBar.xaml.cs (hunk 7 above)
Comment (yaira2) [reply src/Files.App/UserControls/TabBar/TabBar.xaml.cs:388]: I would prefer if this is reverted

**Review by yaira2 — APPROVED** [review]
LGTM

**Comment by XTorLukas** [comment]
Solved problems in this revert PR #16595

**Pull request merged**